ray-operator/controllers/raycluster_controller.go

```diff
@@ -248,6 +248,35 @@ func (r *RayClusterReconciler) reconcilePods(instance *rayiov1alpha1.RayCluster)
 			}
 		}
 		diff := *worker.Replicas - int32(len(runningPods.Items))
+
+		//// SriramQ: How do I create a feature flag to guard the new functionality?
+		featureFlag := true
+
+		if featureFlag {
+			// Always remove the specified WorkersToDelete - regardless of the value of Replicas.
+			// Essentially WorkersToDelete has to be deleted to meet the expectations of the Autoscaler.
+			log.Info("reconcilePods", "removing all the pods in the scaleStrategy of", worker.GroupName)
+			for _, podsToDelete := range worker.ScaleStrategy.WorkersToDelete {
+				pod := corev1.Pod{}
+				pod.Name = podsToDelete
+				pod.Namespace = utils.GetNamespace(instance.ObjectMeta)
+				log.Info("Deleting pod", "namespace", pod.Namespace, "name", pod.Name)
+				if err := r.Delete(context.TODO(), &pod); err != nil {
```

Comment:
Maybe firstly we should check the replica diff before delete pod?

Reply:
I feel the strategy is a little bit different and we should have consensus on the expected behavior. 

Let's use this concrete case as an example.   e.g. expected > running with non-empty WorkersToDelete

Option 1 (chenk008): The configuration is incorrect, then we should ignore it or alert the user.

Option 2 (sriram-anyscale): Above case is valid, user expected to delete `WorkersToDelete` first, and then bump workers to expected number instead of reusing pods from `WorkersToDelete`.

Reply:
I think `Option 2` looks like recreating some pods.

Reply:
1. `The Ray cluster really does not expect these nodes to be running any more - it expects brand new nodes to be added in the case of "expected > running".` you should delete these pods directly, the operator will create new pods. It is not declarative that Use `WorkersToDelete` to recreate pod .
2. `50 nodes die` doesn't mean these pod will lost, these node will restart and join the raycluster.
3. `half the entries in WorkersToDelete being invalid` is incorrect, they are all valid pod to operator.

Reply:
I agree that we should delete these pods directly. As I have commented earlier, this is not possible unfortunately given the current Kuberay design, and furthermore we are already doing this. This issue is not something this PR introduces. It just improves handling given the current constraints.

On your second comment - "doesn't mean these pod will lost, these node will restart and join the raycluster" that is correct, but that will only happen in this reconciliation loop. It is the code here that restarts these nodes.

On the third comment, my use of "invalid" is that these nodes are not running.

Reply:
> If you look at the Autoscaler code, it has already asked the GCS to drain these nodes

@sriram-anyscale 
That was probably a design error. It might have been better to do the draining via preStop hook. In fact, there's a bit of redundancy right now -- the example configs include `ray stop` as a preStop hook.
Currently, `ray stop` and the drain request behave in very similar ways.

The implementation of "drain" is pretty much just "kill" -- though we're thinking about changing that in the future.

The fact that the autoscaler asks the GCS to drain nodes is a problem -- the autoscaler (at least in my opinion) is supposed to produce scale requests, not do node management.

Reply:
> The implementation of "drain" is pretty much just "kill" -- though we're thinking about changing that in the future.

Referring to the future design of draining-  would it make sense to allow certain nodes to finish executing all tasks submitted and then not include them in subsequent scheduling cycles inside the ray core?

Reply:
Yep, that's what we're thinking.

Reply:
@asm582 it might make sense to open an issue for that in the ray repo so we don't forget.

Reply:
Re: main thread of this, Option 2 seems pretty logically coherent to me.
"Achieve a scale of N workers, in the process making sure workers [x,y,z] are gone."
Whether `N <= current - len(workersToDelete)` shouldn't make a difference.

Reply:
> @asm582 it might make sense to open an issue for that in the ray repo so we don't forget.

Done https://github.com/ray-project/ray/issues/23522